### PR TITLE
Add tests for BEFORE/AFTER block

### LINKED ISSUE #946: Implement BEFORE / AFTER blocks
Check the `BEFORE` and `AFTER` blocks implementation and documentation in `NEURON` and add related unit test in `NMODL`.

### PR DESCRIPTION
**Pull request by alkino**
Should fix #946

### CODE REVIEW

File: test/unit/codegen/codegen_cpp_visitor.cpp [hunk 1]
```diff
@@ -422,3 +422,60 @@ SCENARIO("Check code generation for TABLE statements", "[codegen][array_variable
...
     }
 }
+
+SCENARIO("Check that BEFORE/AFTER block are well generated", "[codegen][before/after]") {
+    GIVEN("A mod file full of BEFORE/AFTER of all kinds") {
+        std::string const nmodl_text = R"(
+            NEURON {
+                SUFFIX ba1
+            }
+            BEFORE BREAKPOINT {}
```
Comment (ohm314) [inline test/unit/codegen/codegen_cpp_visitor.cpp:432]: Wondering if we should check how it looks like with something inside the before/after block?

**Comment by ohm314** [comment]
@alkino did you see the spack failure on the bbpgitlab CI? Maybe it has something to do with the updated neuron@develop that was mentioned today? I think once this hickup is fixed and maybe the tests are just a little bit expanded we can merge this.

File: test/unit/codegen/codegen_cpp_visitor.cpp (hunk 1 above)
Comment (iomaganaris) [reply test/unit/codegen/codegen_cpp_visitor.cpp:432]: I agree with Omar, this would be nice to have

**Comment by alkino** [comment]
The generated code is verbose.

```
BEFORE INITIAL {
  init()
  inc = 0
  bi = inc
  bit = t
  printf("ba0 BEFORE INITIAL inc=%g t=%g\n", inc, t)
}
```

generate:

```c++
    /** BEFORE of block type INITIAL # 0 */
    void nrn_before_after_0_ba0(NrnThread* nt, Memb_list* ml, int type) {
        int nodecount = ml->nodecount;
        int pnodecount = ml->_nodecount_padded;
        const int* node_index = ml->nodeindices;
        double* data = ml->data;
        const double* voltage = nt->_actual_v;
        Datum* indexes = ml->pdata;
        ThreadDatum* thread = ml->_thread;
        auto* const inst = static_cast<ba0_Instance*>(ml->instance);

        #pragma ivdep
        #pragma omp simd
        for (int id = 0; id < nodecount; id++) {
            int node_id = node_index[id];
            double v = voltage[node_id];
            #if NRN_PRCELLSTATE
            inst->v_unused[id] = v;
            #endif
            {
                init_ba0(id, pnodecount, inst, data, indexes, thread, nt, v);
                inst->inc[id] = 0.0;
                inst->bi[id] = inst->inc[id];
                inst->bit[id] = nt->_t;
                printf("ba0 BEFORE INITIAL inc=%g t=%g\n", inst->inc[id], nt->_t);
            }
        }
    }
```

I'm not sure what to test there

**Comment by ohm314** [comment]
Maybe it's enough to have 

```
BEFORE INITIAL {
  init()
  inc = 0
}
```

and make sure we find this loop in the output:

```c++
        #pragma ivdep
        #pragma omp simd
        for (int id = 0; id < nodecount; id++) {
            int node_id = node_index[id];
            double v = voltage[node_id];
            #if NRN_PRCELLSTATE
            inst->v_unused[id] = v;
            #endif
            {
                init_ba0(id, pnodecount, inst, data, indexes, thread, nt, v);
                inst->inc[id] = 0.0;
            }
        }
```

File: test/unit/codegen/codegen_cpp_visitor.cpp [hunk 2]
```diff
@@ -423,6 +423,163 @@ SCENARIO("Check code generation for TABLE statements", "[codegen][array_variable
...
+        )";
+        THEN("They should be all registered") {
+            auto const generated = get_cpp_code(nmodl_text);
+            REQUIRE_THAT(generated, Contains("hoc_reg_ba(mech_type, nrn_before_after_0_ba1, 14);"));
+            REQUIRE_THAT(generated, Contains("hoc_reg_ba(mech_type, nrn_before_after_1_ba1, 22);"));
+            REQUIRE_THAT(generated, Contains("hoc_reg_ba(mech_type, nrn_before_after_2_ba1, 14);"));
+            REQUIRE_THAT(generated, Contains("hoc_reg_ba(mech_type, nrn_before_after_3_ba1, 22);"));
+        }
+    }
+}
```
Comment (pramodk) [inline test/unit/codegen/codegen_cpp_visitor.cpp:581]: nice 👌

**Review by pramodk — APPROVED** [review]
👍 thanks!

If CI passes, good to go!

**Review by Yedfree — APPROVED** [review]

**Pull request merged**